File: surf-friends/src/main/java/dev/slne/surf/friends/menu/sub/friend/FriendFriendsMenu.java
package dev.slne.surf.friends.menu.sub.friend;

import com.github.stefvanschie.inventoryframework.pane.OutlinePane;
import com.github.stefvanschie.inventoryframework.pane.PaginatedPane;
import com.github.stefvanschie.inventoryframework.pane.Pane.Priority;
import com.github.stefvanschie.inventoryframework.pane.StaticPane;
import dev.slne.surf.friends.FriendManager;
import dev.slne.surf.friends.core.util.FriendLogger;
import dev.slne.surf.friends.core.util.ItemBuilder;
import dev.slne.surf.friends.core.util.PluginColor;
import dev.slne.surf.friends.menu.FriendMainMenu;
import dev.slne.surf.friends.menu.FriendMenu;
import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import it.unimi.dsi.fastutil.objects.ObjectList;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;


public class FriendFriendsMenu extends FriendMenu {
  public FriendFriendsMenu(UUID player) {
    super(5, "Deine Freunde");

    OutlinePane header = new OutlinePane(0, 0, 9, 1, Priority.LOW);
    OutlinePane footer = new OutlinePane(0, 4, 9, 1, Priority.LOW);
    PaginatedPane pages = new PaginatedPane(1, 1, 9, 3, Priority.HIGH);
    StaticPane navigation = new StaticPane(0, 4, 9, 1, Priority.HIGH);

    header.addItem(build(new ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).setName("")));
    header.setRepeat(true);

    footer.addItem(build(new ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).setName("")));
    footer.setRepeat(true);

    pages.populateWithItemStacks(this.getFriendItems(player));

    pages.setOnClick(event -> {
      if(event.getCurrentItem() == null){
        return;
      }

      ItemMeta meta = event.getCurrentItem().getItemMeta();

      new FriendSingleMenu(meta.getDisplayName()).show(event.getWhoClicked());
    });

    navigation.addItem(build(new ItemBuilder(Material.RED_WOOL).setName(Component.text("Vorherige Seite").color(PluginColor.RED)), event -> {
      if (pages.getPage() > 0) {
        pages.setPage(pages.getPage() - 1);

        update();
      }
    }), 0, 0);

    navigation.addItem(build(new ItemBuilder(Material.GREEN_WOOL).setName(Component.text("Nächste Seite").color(PluginColor.LIGHT_GREEN)), event -> {
      if (pages.getPage() < pages.getPages() - 1) {
        pages.setPage(pages.getPage() + 1);
        update();
      }
    }), 8, 0);

    navigation.addItem(build(new ItemBuilder(Material.BARRIER).setName(Component.text("Zurück").color(PluginColor.RED)), event -> new FriendMainMenu().show(event.getWhoClicked())), 4, 0);


    addPane(header);
    addPane(footer);
    addPane(navigation);
    addPane(pages);


    setOnGlobalClick(event -> event.setCancelled(true));
    setOnGlobalDrag(event -> event.setCancelled(true));
  }


  private ObjectList<ItemStack> getFriendItems(UUID player){
    ObjectList<ItemStack> stacks = new ObjectArrayList<>();

    FriendManager.instance().getFriends(player).thenAccept(friends -> {
      for (UUID friend : friends) {
        OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(friend);

        stacks.add(new ItemBuilder(Material.PLAYER_HEAD).setName(offlinePlayer.getName()).setSkullOwner(offlinePlayer.getName()).build());
      }
    });

    return stacks;
  }
}

File: surf-friends/src/main/java/dev/slne/surf/friends/menu/sub/friend/FriendFriendsMenu.kt
package dev.slne.surf.friends.menu.sub.friend

import com.github.stefvanschie.inventoryframework.font.util.Font
import com.github.stefvanschie.inventoryframework.gui.GuiItem
import com.github.stefvanschie.inventoryframework.pane.OutlinePane
import com.github.stefvanschie.inventoryframework.pane.PaginatedPane
import com.github.stefvanschie.inventoryframework.pane.Pane
import com.github.stefvanschie.inventoryframework.pane.StaticPane
import com.github.stefvanschie.inventoryframework.pane.component.Label
import dev.slne.surf.friends.listener.util.ItemBuilder
import dev.slne.surf.friends.listener.util.PluginColor
import dev.slne.surf.friends.listener.util.buildGuiItem
import dev.slne.surf.friends.menu.FriendMainMenu
import dev.slne.surf.friends.menu.FriendMenu
import dev.slne.surf.friends.menu.backItem
import dev.slne.surf.friends.menu.buttons.FriendButton
import dev.slne.surf.friends.menu.getPagesButtons
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import it.unimi.dsi.fastutil.objects.ObjectList
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.*

class FriendFriendsMenu(friends: ObjectList<UUID>) : FriendMenu(5, "Deine Freunde") {
    init {
        val header = OutlinePane(0, 0, 9, 1, Pane.Priority.LOW)
        val footer = OutlinePane(0, 4, 9, 1, Pane.Priority.LOW)
        val pages = PaginatedPane(1, 1, 9, 3, Pane.Priority.HIGH)
        val navigation = StaticPane(0, 4, 9, 1, Pane.Priority.HIGH)
        val addFriend = Label(4, 0, 1, 1, Pane.Priority.NORMAL, Font.OAK_PLANKS)

        header.addItem(build(ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).setName("")))
        header.setRepeat(true)

        footer.addItem(build(ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).setName("")))
        footer.setRepeat(true)

        addFriend.setText("+") { _: Char?, stack: ItemStack ->
            val builder = ItemBuilder(stack)

            builder.setName(Component.text("Freund/in hinzufügen"))
            builder.addLoreLine(Component.text("Freund/in hinzufügen", PluginColor.DARK_GRAY).decoration(TextDecoration.ITALIC, false))
            builder.addLoreLine(Component.text(" ", PluginColor.LIGHT_GRAY).decoration(TextDecoration.ITALIC, false))
            builder.addLoreLine(Component.text("Hier kannst du eine Person", PluginColor.LIGHT_GRAY).decoration(TextDecoration.ITALIC, false))
            builder.addLoreLine(Component.text("zu deiner Freundschaftsliste", PluginColor.LIGHT_GRAY).decoration(TextDecoration.ITALIC, false))
            builder.addLoreLine(Component.text("hinzufügen.", PluginColor.LIGHT_GRAY).decoration(TextDecoration.ITALIC, false))

            this.build(builder) {
                if (it == null) {
                    return@build
                }

                FriendAddMenu().open(it.whoClicked as Player)
            }
        }

        pages.populateWithGuiItems(getFriendItems(friends))
        navigation.addItem(buildGuiItem(backItem) { event -> FriendMainMenu().show(event.whoClicked) }, 4, 0)

        addPane(header)
        addPane(footer)
        addPane(navigation)
        addPane(getPagesButtons(pages))
        addPane(pages)
        addPane(addFriend)

        setOnGlobalClick {
            it.isCancelled = true
        }

        setOnGlobalDrag {
            it.isCancelled = true
        }
    }

    private fun getFriendItems(friends: ObjectList<UUID>): ObjectList<GuiItem> {
        val stacks: ObjectList<GuiItem> = ObjectArrayList()

        friends.forEach {
            stacks.add(FriendButton(Bukkit.getOfflinePlayer(it)))
        }

        return stacks
    }
}
